edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent/MetricsScraper.cs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Net;
+    using System.Net.Http;
+    using System.Text.RegularExpressions;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class MetricsScraper : IMetricsScraper, IDisposable
+    {
+        const string UrlPattern = @"[^/:]+://(?<host>[^/:]+)(:[^:]+)?$";
+        static readonly Regex UrlRegex = new Regex(UrlPattern, RegexOptions.Compiled);
+        readonly HttpClient httpClient;
+        readonly Lazy<IDictionary<string, string>> endpoints;
+        static readonly ILogger Log = Logger.Factory.CreateLogger<MetricsScraper>();
+
+        public MetricsScraper(IList<string> endpoints)
+        {
+            this.httpClient = new HttpClient();
+            this.endpoints = new Lazy<IDictionary<string, string>>(() => endpoints.ToDictionary(e => e, this.GetUriWithIpAddress));
```

Comment:
why use lazy? is it time-consuming?

Reply:
for some reason the dns service doesn't work immediately on startup. This delays the call until it's needed.